tests/ut/azure_iot_json_reader_ut.c

```diff
@@ -0,0 +1,582 @@
+
+
+
+uint32_t ulGetAllTests();
+
+static void testAzureIoTJSONReader_Init_Failure( void ** ppvState )
+{
+    AzureIoTJSONReader_t xReader;
+    uint8_t * pucTestJSON = azureiotjsonreaderutTEST_JSON_STRING;
+
+    /* Fail init if JSON reader is NULL */
+    assert_int_equal( AzureIoTJSONReader_Init( NULL,
+                                               pucTestJSON,
+                                               strlen( pucTestJSON ) ),
+                      eAzureIoTErrorInvalidArgument );
+
+    /* Fail init if JSON text is NULL */
+    assert_int_equal( AzureIoTJSONReader_Init( &xReader,
+                                               NULL,
+                                               strlen( pucTestJSON ) ),
+                      eAzureIoTErrorInvalidArgument );
+
+    /* Fail init if JSON text length is 0 */
+    assert_int_equal( AzureIoTJSONReader_Init( &xReader,
+                                               pucTestJSON,
+                                               0 ),
+                      eAzureIoTErrorInvalidArgument );
+}
+
+static void testAzureIoTJSONReader_Init_Success( void ** ppvState )
+{
+    AzureIoTJSONReader_t xReader;
+    uint8_t * pucTestJSON = azureiotjsonreaderutTEST_JSON;
+
+    assert_int_equal( AzureIoTJSONReader_Init( &xReader,
+                                               pucTestJSON,
+                                               strlen( pucTestJSON ) ),
+                      eAzureIoTSuccess );
+}
+
+static void testAzureIoTJSONReader_NextToken_Failure( void ** ppvState )
+{
+    /* Fail init if JSON reader is NULL */
+    assert_int_equal( AzureIoTJSONReader_NextToken( NULL ),
+                      eAzureIoTErrorInvalidArgument );
+}
+
+static void testAzureIoTJSONReader_NextToken_Success( void ** ppvState )
+{
+    AzureIoTJSONReader_t xReader;
+    AzureIoTJSONTokenType_t xTokenType;
+
+    assert_int_equal( AzureIoTJSONReader_Init( &xReader, azureiotjsonreaderutTEST_JSON, strlen( azureiotjsonreaderutTEST_JSON ) ),
+                      eAzureIoTSuccess );
+
+    /*Begin object */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenBEGIN_OBJECT );
+
+    /*Property name */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenPROPERTY_NAME );
+
+    /*Property string */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenSTRING );
```

Comment:
>eAzureIoTJSONTokenSTRING [](http://example.com/codeflow?start=34&length=24)

Can we build one JSON that has all the types we support and then read that ..
Also having invalid json read also can be good test case, but having fuz test for that can be better approach

[https://expresslogic.visualstudio.com/X-Ware/_git/netx?path=%2Fnx_secure_internal%2FTest%2FFuzz%2Fazure_iot%2Fazure_iot_process_json_reader_test.c&version=GBmaster&_a=contents](https://expresslogic.visualstudio.com/X-Ware/_git/netx?path=%2Fnx_secure_internal%2FTest%2FFuzz%2Fazure_iot%2Fazure_iot_process_json_reader_test.c&version=GBmaster&_a=contents)

Reply:
In theory we could but at this point, is it worth the code rework or is what's here sufficient? Essentially is it worth the time if everything is checked already.